### PR TITLE
Add tenantId and scopes to TokenCacheNotificationArgs

### LINKED ISSUE #3389: [Feature Request] Expose resource (scopes) in TokenCacheNotificationArgs
**Is your feature request related to a problem? Please describe.**
#3385 As described here, now the app token is partitioned by tenantid, when using it with single-tenant, it will become a memory issue with large resource amount.  And if we want to enable distributed cache, the value's size is a concern, because the suggestedkey is {clientId}_{tenantId}_AppTokenCache, so all the access tokens will be serialized into a byte array, this will cause a LOH allocation and a large redis vaule (if we use redis as distributed cache).

**Describe the solution you'd like**
1. Expose the resource (scope) in TokenCacheNotificationArgs.
~2. Change the serializer behavior to support single access token serializations.~

### PR DESCRIPTION
**Pull request by bgavrilMS**
Fixes #3389 

**Changes proposed in this request**
Add new properties to TokenCacheNotificationArgs

**Testing**
unit + int

**Performance impact**
no operation except reading existing properties.

### CODE REVIEW

File: tests/Microsoft.Identity.Test.Unit/Microsoft.Identity.Test.Unit.csproj [hunk 1]
```diff
@@ -50,7 +50,6 @@
     <ProjectReference Include="..\..\src\client\Microsoft.Identity.Client.Desktop\Microsoft.Identity.Client.Desktop.csproj" />
   </ItemGroup>
   <ItemGroup Condition="'$(TargetFramework)' == 'netcoreapp3.1' ">
-    <PackageReference Include="Microsoft.Windows.SDK.Contracts" Version="10.0.19041.1" />
```
Comment (bgavrilMS) [inline tests/Microsoft.Identity.Test.Unit/Microsoft.Identity.Test.Unit.csproj:53]: unrelated change: eliminate a warning.

File: tests/Microsoft.Identity.Test.Unit/PublicApiTests/ConfidentialClientApplicationTests.cs [hunk 2]
```diff
@@ -1681,7 +1677,7 @@ public async Task ValidateAppTokenProviderAsync()
                                                               {
                                                                   Assert.IsNotNull(parameters.Claims);
                                                               }
-                                                              return Task.Run(() => GetAppTokenProviderResult(differentScopesForAt)).Result;
+                                                              return await Task.Run(() => GetAppTokenProviderResult(differentScopesForAt)).ConfigureAwait(false);
```
Comment (bgavrilMS) [inline tests/Microsoft.Identity.Test.Unit/PublicApiTests/ConfidentialClientApplicationTests.cs:1680]: unrelated change: eliminate a warning.

File: tests/Microsoft.Identity.Test.Unit/pop/PoPTests.cs [hunk 3]
```diff
@@ -2,15 +2,12 @@
...
 
 using System;
-using System.Collections.Generic;
 using System.Diagnostics;
 using System.Globalization;
 using System.Linq;
 using System.Net.Http;
 using System.Security.Cryptography;
 using System.Text;
-using System.Linq;
```
Comment (bgavrilMS) [inline tests/Microsoft.Identity.Test.Unit/pop/PoPTests.cs:12]: unrelated change: eliminate a warning.

File: src/client/Microsoft.Identity.Client/TokenCache.ITokenCacheInternal.cs [hunk 4]
```diff
@@ -232,7 +234,10 @@ public sealed partial class TokenCache : ITokenCacheInternal
                             hasTokens: tokenCacheInternal.HasTokensNoLocks(),
                             suggestedCacheExpiry: cacheExpiry,
                             cancellationToken: requestParams.RequestContext.UserCancellationToken,
-                            correlationId: requestParams.RequestContext.CorrelationId);
+                            correlationId: requestParams.RequestContext.CorrelationId, 
+                            requestScopes: requestParams.Scope,
+                            requestTenantId: requestParams.AuthorityManager.OriginalAuthority.TenantId);
+
 
                         Stopwatch sw = Stopwatch.StartNew();
```
Comment (trwalke) [inline src/client/Microsoft.Identity.Client/TokenCache.ITokenCacheInternal.cs:242]: nit: Extra line above

File: src/client/Microsoft.Identity.Client/TokenCache.ITokenCacheInternal.cs [hunk 5]
```diff
@@ -1168,7 +1178,10 @@ async Task ITokenCacheInternal.RemoveAccountAsync(IAccount account, Authenticati
                            hasTokens: tokenCacheInternal.HasTokensNoLocks(),
                            suggestedCacheExpiry: null,
                            cancellationToken: requestParameters.RequestContext.UserCancellationToken,
-                           correlationId: requestParameters.RequestContext.CorrelationId);
+                           correlationId: requestParameters.RequestContext.CorrelationId,
+                           requestScopes: requestParameters.Scope,
+                           requestTenantId: requestParameters.AuthorityManager.OriginalAuthority.TenantId);
+
 
                         await tokenCacheInternal.OnAfterAccessAsync(args).ConfigureAwait(false);
```
Comment (trwalke) [inline src/client/Microsoft.Identity.Client/TokenCache.ITokenCacheInternal.cs:1186]: nit: Extra line above

File: src/client/Microsoft.Identity.Client/AppConfig/AssertionRequestOptions.cs [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+        /// <summary>
+        /// Information about the client assertion that need to be generated See https://aka.ms/msal-net-client-assertion
+        /// </summary>
+        /// <remarks> Use the provided information to generate the client assertion payload </remarks>
+#if !SUPPORTS_CONFIDENTIAL_CLIENT
+    [System.ComponentModel.EditorBrowsable(System.ComponentModel.EditorBrowsableState.Never)]  // hide confidential client on mobile
+#endif
+    public class AssertionRequestOptions {
+        /// <summary>
+        /// </summary>
```
Comment (SameerK-MSFT) [inline src/client/Microsoft.Identity.Client/AppConfig/AssertionRequestOptions.cs:13]: nit: you can remove the empty comment (or add summary to it)

**Comment by bgavrilMS** [comment]
Got into merge hell with this one, closing. Will use a copy - https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/3392

File: src/client/Microsoft.Identity.Client/TokenCacheNotificationArgs.cs [hunk 7]
```diff
@@ -53,7 +57,39 @@ public sealed partial class TokenCacheNotificationArgs
...
+            bool hasStateChanged,
+            bool isApplicationCache,
+            string suggestedCacheKey,
+            bool hasTokens,
+            DateTimeOffset? suggestedCacheExpiry,
+            CancellationToken cancellationToken,
+            Guid correlationId, 
+            IEnumerable<string> requestScopes,
+            string requestTenantId)
+
```
Comment (SameerK-MSFT) [inline src/client/Microsoft.Identity.Client/TokenCacheNotificationArgs.cs:92]: not: extra line

**Review by SameerK-MSFT — APPROVED** [review]

**Review by SameerK-MSFT — COMMENTED** [review]
LGTM. There is no approve button :)
